File: src/fx/fit.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

#include "zeronn/zeronn.h"

#define DROP 0.5
#define MOM (float)(7.0 / 8)
#define EPS (float)(1.0 / 1024)

layer_t* get_block(int in_dim, int out_dim) {
    return sequence(
        relu(),
        dropout(fx_from_fp(DROP)),
        linear(in_dim, out_dim),
        batchnorm0d(out_dim, fx_from_fp(MOM), fx_from_fp(EPS))
    );
}

layer_t* get_model(int in_dim, int out_dim) {
    int a = 64;
    int b = 32;
    int c = 16;
    return sequence(
        linear(in_dim, a),
        batchnorm0d(a, fx_from_fp(MOM), fx_from_fp(EPS)),
        get_block(a, b),
        get_block(b, c),
        relu(),
        linear(c, out_dim),
        softmax()
    );
}

void get_batch(int out_dim, fx_t signal, tensor_t* x, int* y) {
    int batch_size = x->shape[0];
    int in_dim = x->shape[1];

    for (int i = 0; i < batch_size; ++i) {
        y[i] = prng_randint(0, out_dim);
    }

    tensor_uniform(x, 0, FX_ONE - signal);

    assert(in_dim % out_dim == 0);
    for (int i = 0; i < batch_size; ++i) {
        for (int j = y[i]; j < in_dim; j += out_dim) {
            x->data[i * in_dim + j] += signal;
        }
    }
}

tensor_t* get_loss(tensor_t* y_pred, int* y_gold) {
    int batch_size = y_pred->shape[0];
    int dim = y_pred->shape[1];
    tensor_t* dy = tensor_clone(y_pred);
    for (int i = 0; i < batch_size; ++i) {
        dy->data[i * dim + y_gold[i]] -= FX_ONE;
    }
    for (int i = 0; i < dy->size; ++i) {
        dy->data[i] = fx_div(dy->data[i], dim);
    }
    return dy;
}

int get_correct(tensor_t* y_pred, int* y_gold) {
    int batch_size = y_pred->shape[0];
    int dim = y_pred->shape[1];
    int r = 0;
    for (int i = 0; i < batch_size; ++i) {
        int j_max = 0;
        fx_t y_max = y_pred->data[i * dim];
        for (int j = 1; j < dim; ++j) {
            fx_t y = y_pred->data[i * dim + j];
            if (y_max < y) {
                j_max = j;
                y_max = y;
            }
        }
        if (j_max == y_gold[i]) {
            ++r;
        }
    }
    return r;
}

int main(void) {
    int num_epochs = 10;
    int rounds_per_epoch = 20;
    int train_per_round = 5;
    int eval_per_round = 1;
    int batch_size = 256;
    int in_dim = 128;
    int out_dim = 8;
    fx_t signal = fx_from_fp(0.25);
    fx_t lr = fx_from_fp(1.0 / 512);

    layer_t* f = get_model(in_dim, out_dim);

    tensor_t* x = zeros(batch_size, in_dim);
    int* y_gold = (int*)malloc(batch_size * sizeof(int));

    for (int i = 0; i < num_epochs; ++i) {
        int ta = 0;
        int va = 0;
        for (int j = 0; j < rounds_per_epoch; ++j) {
            for (int k = 0; k < train_per_round; ++k) {
                get_batch(out_dim, signal, x, y_gold);
                layer_zero_grad(f);
                tensor_t* y_pred = layer_forward(f, x, true);
                tensor_t* dy = get_loss(y_pred, y_gold);
                tensor_t* dx = layer_backward(f, dy);
                layer_update_step(f, lr);
                ta += get_correct(y_pred, y_gold);
                tensor_free(dx);
                free(dx);
                tensor_free(dy);
                free(dy);
                tensor_free(y_pred);
                free(y_pred);
            }
            for (int k = 0; k < eval_per_round; ++k) {
                get_batch(out_dim, signal, x, y_gold);
                tensor_t* y_pred = layer_forward(f, x, true);
                va += get_correct(y_pred, y_gold);
                tensor_free(y_pred);
                free(y_pred);
            }
        }
        int t = (ta * 10000) / (rounds_per_epoch * train_per_round *
                                batch_size);
        int v = (va * 10000) / (rounds_per_epoch * eval_per_round *
                                batch_size);
        printf("%7d %3d.%03d %3d.%03d\n", i, t / 100, t % 100, v / 100,
               v % 100);
    }

    free(y_gold);
    tensor_free(x);
    free(x);
    layer_free(f);
    free(f);
}

File: zeronn/fx/fit.py
from argparse import ArgumentParser
import torch

from .zeronn import *


def parse_args():
    x = ArgumentParser()
    x.add_argument('--num_epochs', type=int, default=10)
    x.add_argument('--rounds_per_epoch', type=int, default=20)
    x.add_argument('--train_per_round', type=int, default=5)
    x.add_argument('--eval_per_round', type=int, default=1)
    x.add_argument('--batch_size', type=int, default=32)
    x.add_argument('--in_dim', type=int, default=128)
    x.add_argument('--out_dim', type=int, default=8)
    x.add_argument('--signal', type=float, default=0.25)
    x.add_argument('--lr', type=float, default=1 / 128)
    return x.parse_args()


def get_block(in_dim, out_dim):
    return Sequence(
        ReLU(),
        Dropout(),
        Linear(in_dim, out_dim),
        BatchNorm0d(out_dim),
    )


def get_model(in_dim, out_dim):
    a = 64
    b = 32
    c = 16
    return Sequence(
        Linear(in_dim, a),
        BatchNorm0d(a),
        get_block(a, b),
        get_block(b, c),
        ReLU(),
        Linear(c, out_dim),
        Softmax(),
    )


def make_batch(batch_size, in_dim, out_dim, signal):
    assert out_dim <= in_dim
    assert in_dim % out_dim == 0
    num_repeats = in_dim // out_dim
    y = torch.randint(0, out_dim, (batch_size,))
    y_one_hot = torch.eye(out_dim)[y]
    r = torch.rand(batch_size, num_repeats, out_dim)
    x = signal * y_one_hot.unsqueeze(1) + (1 - signal) * r
    x = fx_from_fp(x)
    return x.view(batch_size, -1), y


def main(args):
    f = get_model(args.in_dim, args.out_dim)
    get_batch = lambda: make_batch(args.batch_size, args.in_dim, args.out_dim,
                                   args.signal)
    for epoch_id in range(args.num_epochs):
        taa = []
        vaa = []
        for round_id in range(args.rounds_per_epoch):
            for batch_id in range(args.train_per_round):
                f.zero_grad()
                x, y_gold = get_batch()
                y_pred = f(x, True)
                dy = y_pred.clone()
                dy[torch.arange(args.batch_size), y_gold] -= fx_one
                dy = fx_div(dy, args.out_dim)
                f.backward(dy)
                f.update_step(args.lr)
                a = (y_pred.argmax(1) == y_gold).float().mean()
                taa.append(a.item())
            with torch.no_grad():
                for batch_id in range(args.eval_per_round):
                    x, y_gold = get_batch()
                    y_pred = f(x, False)
                    a = (y_pred.argmax(1) == y_gold).float().mean()
                    vaa.append(a.item())
        ta = sum(taa) / len(taa)
        va = sum(vaa) / len(vaa)
        s = '%7d %7.3f %7.3f' % (epoch_id, 100 * ta, 100 * va)
        print(s)


if __name__ == '__main__':
    main(parse_args())
